deepmd/pt/model/descriptor/repformer_layer.py

```diff
@@ -442,154 +898,48 @@ def _update_g1_conv(
         if not self.smooth:
             # normalized by number of neighbors, not smooth
             # nb x nloc x 1
-            invnnei = 1.0 / (self.epsilon + torch.sum(nlist_mask, dim=-1)).unsqueeze(-1)
+            # must use type_as here to convert bool to float, otherwise there will be numerical difference from numpy
+            invnnei = 1.0 / (
+                self.epsilon + torch.sum(nlist_mask.type_as(gg1), dim=-1)
+            ).unsqueeze(-1)
         else:
             gg1 = _apply_switch(gg1, sw)
             invnnei = (1.0 / float(nnei)) * torch.ones(
-                (nb, nloc, 1), dtype=env.GLOBAL_PT_FLOAT_PRECISION, device=gg1.device
+                (nb, nloc, 1), dtype=gg1.dtype, device=gg1.device
             )
         # nb x nloc x ng2
         g1_11 = torch.sum(g2 * gg1, dim=2) * invnnei
         return g1_11
 
-    def _cal_h2g2(
-        self,
-        g2: torch.Tensor,
-        h2: torch.Tensor,
-        nlist_mask: torch.Tensor,
-        sw: torch.Tensor,
-    ) -> torch.Tensor:
-        # g2:  nf x nloc x nnei x ng2
-        # h2:  nf x nloc x nnei x 3
-        # msk: nf x nloc x nnei
-        nb, nloc, nnei, _ = g2.shape
-        ng2 = g2.shape[-1]
-        # nb x nloc x nnei x ng2
-        g2 = _apply_nlist_mask(g2, nlist_mask)
-        if not self.smooth:
-            # nb x nloc
-            invnnei = 1.0 / (self.epsilon + torch.sum(nlist_mask, dim=-1))
-            # nb x nloc x 1 x 1
-            invnnei = invnnei.unsqueeze(-1).unsqueeze(-1)
-        else:
-            g2 = _apply_switch(g2, sw)
-            invnnei = (1.0 / float(nnei)) * torch.ones(
-                (nb, nloc, 1, 1), dtype=env.GLOBAL_PT_FLOAT_PRECISION, device=g2.device
-            )
-        # nb x nloc x 3 x ng2
-        h2g2 = torch.matmul(torch.transpose(h2, -1, -2), g2) * invnnei
-        return h2g2
-
-    def _cal_grrg(self, h2g2: torch.Tensor) -> torch.Tensor:
-        # nb x nloc x 3 x ng2
-        nb, nloc, _, ng2 = h2g2.shape
-        # nb x nloc x 3 x axis
-        h2g2m = torch.split(h2g2, self.axis_dim, dim=-1)[0]
-        # nb x nloc x axis x ng2
-        g1_13 = torch.matmul(torch.transpose(h2g2m, -1, -2), h2g2) / (3.0**1)
-        # nb x nloc x (axisxng2)
-        g1_13 = g1_13.view(nb, nloc, self.axis_dim * ng2)
-        return g1_13
-
-    def _update_g1_grrg(
```

Comment:
do not mv the position of grrg in this pr.
you may want to reused the symmetry op in all dp descriptors. refact it in a future PR.